crates/conf/macros/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+//! This crate provides a derive macro that provides the necessary functionality for a struct to be
+//! used as a configuration object and be bundled into.
+
+use itertools::Itertools;
+use proc_macro::TokenStream;
+use proc_macro2::Span;
+use quote::{quote, ToTokens};
+use syn::parse::{Parse, ParseStream};
+use syn::{parse_macro_input, Data, DeriveInput, Fields, Pat, Token};
+
+struct EnsureArgs(syn::Expr, syn::LitStr);
+impl Parse for EnsureArgs {
+    fn parse(input: ParseStream) -> Result<EnsureArgs, syn::Error> {
+        let condition = input.parse()?;
+        input.parse::<Token![,]>()?;
+        let string = input.parse()?;
+        Ok(EnsureArgs(condition, string))
+    }
+}
+
+#[derive(Clone)]
+struct ConfigField {
+    check_fns: Vec<Option<proc_macro2::TokenStream>>,
+    field_name: syn::Ident,
+    field_type: syn::Type,
+    field_publicity: syn::Visibility,
+}
+
+#[proc_macro_derive(Config, attributes(check, ensure, is_finite))]
+pub fn derive_config(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let struct_name = &input.ident;
+    let data = &input.data;
+
+    let mut config_fields: Vec<ConfigField> = Vec::new();
+
+    let struct_publicity = &input.vis;
+    let mut is_finite_check = false;
+
+    if let Data::Struct(data_struct) = data {
+        if let Fields::Named(fields) = &data_struct.fields {
+            for field in &fields.named {
+                let mut field_info = ConfigField {
+                    check_fns: Vec::new(),
+                    field_name: field.ident.clone().unwrap(),
+                    field_type: field.ty.clone(),
+                    field_publicity: field.vis.clone(),
+                };
+
+                let field_name = &field_info.field_name;
+                let field_type = &field_info.field_type;
+
+                let inner_check_fns = &mut field_info.check_fns;
+
+                let mut check_fn;
+                for attr in &field.attrs {
+                    // get the check attribute with closures like this: #[check(|v| v > 0)]
+                    if attr.path.is_ident("check") {
+                        let closure = match attr.parse_args::<syn::ExprClosure>() {
```

Comment:
I would separate each branch (`check`, `ensure` and `is_finite`) to a function so that the high level structure of this macro is easier to read.